### PR TITLE
Added isCompositeNumber.js

### PR DESCRIPTION
**Pull request by tehliang**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by appgurueu — CHANGES_REQUESTED** [review]
~~Isn't this the same as the already implemented check for ["perfect numbers"](https://the-algorithms.com/de/algorithm/perfect-number?lang=javascript)?~~

NVM, should've read carefully.

**Comment by tehliang** [comment]
> Isn't this the same as the already implemented check for ["perfect numbers"](https://the-algorithms.com/de/algorithm/perfect-number?lang=javascript)?

No, they are not the same.
From Wikipedia and the internet.
A perfect number is a positive integer equal to the sum of its positive divisors, excluding the number itself. 
For instance, 6 has divisors 1, 2 and 3 (excluding itself), and 1 + 2 + 3 = 6, so 6 is a perfect number.
 
A composite number is a positive integer that can be formed by multiplying two smaller positive integers. Equivalently, it is a positive integer that has at least one divisor other than 1 and itself.
For example, the integer 14 is a composite number because it is the product of the two smaller integers 2 × 7.
 
Example:
35 is a composite number but not a perfect number because 
1 + 5 + 7 not equal to 35
while 35 has more than two factors, i.e. 1, 5, 7, 35.
Which makes 35 a composite number because 35 has more than two factors.

File: Maths/IsCompositeNumber.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * A composite number is a positive integer that is not prime. In other words, it has a positive divisor other than one or itself.
+ * First few composite numbers are 4, 6, 8, 9, 10, 12, 14, 15, 16, 18, 20, ………
+ * Every integer greater than one is either a prime number or a composite number.
+ * The number one is a unit – it is neither prime nor composite.
+ */
+
+function isCompositeNumber (number) {
+  let i = 1
+  let count = 0
+  while (number >= i) {
```
Comment (appgurueu) [inline Maths/IsCompositeNumber.js:17]: Could you reuse one of the existing sieves here?

Comment (tehliang) [reply Maths/IsCompositeNumber.js:17]: > Could you reuse one of the existing sieves here?

Hmm, sorry, I don't get what you mean.

Comment (appgurueu) [reply Maths/IsCompositeNumber.js:17]: There are more efficient ways to factorize a number (such as only searching for factors up to the sqrt). Also, as soon as you have found a single factor, you can divide the number by it to find the second factor (if it exists) & exit early.

Comment (tehliang) [reply Maths/IsCompositeNumber.js:17]: > There are more efficient ways to factorize a number (such as only searching for factors up to the sqrt). Also, as soon as you have found a single factor, you can divide the number by it to find the second factor (if it exists) & exit early.

Oo. I see. I see. Interesting. Do you have any suggestions on how can I improve the algorithm?

Comment (appgurueu) [reply Maths/IsCompositeNumber.js:17]: You could look at using the existing [prime factorization algorithm](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeFactors.js).

Comment (tehliang) [reply Maths/IsCompositeNumber.js:17]: > You could look at using the existing [prime factorization algorithm](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeFactors.js).

Okay, I will try to implement and update the algorithm when I am free. Thanks!

Comment (tehliang) [reply Maths/IsCompositeNumber.js]: Updated the algorithm with a different approach.

**Review by raklaptudirm — APPROVED** [review]

**Review by appgurueu — APPROVED** [review]
Looks good to me. You could perhaps deduplicate the factorization code (if the current APIs aren't too clunky to use).

**Comment by raklaptudirm** [comment]
In my opinion, this is largely a duplication of the [`PrimeCheck`](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeCheck.js#L9) function. Using the fact that all prime numbers other than <samp>2</samp> and <samp>3</samp> are in the form <samp>6n ± 1</samp> is definitely an improvement, but it can be made on the [`PrimeCheck`](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeCheck.js#L9) function itself. @appgurueu?

**Comment by appgurueu** [comment]
> In my opinion, this is largely a duplication of the [`PrimeCheck`](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeCheck.js#L9) function. Using the fact that all prime numbers other than 2 and 3 are in the form 6n ± 1 is definitely an improvement, but it can be made on the [`PrimeCheck`](https://github.com/TheAlgorithms/JavaScript/blob/master/Maths/PrimeCheck.js#L9) function itself. @appgurueu?

Yes, but the `PrimeCheck` function can't be used for this purpose since composite numbers require checking whether a number has at least two factors (e.g. factorizing out at least two factors).

**Comment by raklaptudirm** [comment]
I think instead of duplicating the logic the <samp>6n ± 1</samp> should be made on the <samp>PrimeCheck</samp> function itself.

**Comment by tehliang** [comment]
So any suggestions on how can I improve the algorithm?

**Comment by raklaptudirm** [comment]
You should remove the isComposite function and instead make your changes to the PrimeCheck function.

**Comment by tehliang** [comment]
> You should remove the isComposite function and instead make your changes to the PrimeCheck function.

What changes should I make?

**Comment by tehliang** [comment]
Btw composite numbers up to 100 are:
4, 6, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28, 30, 32, 33, 34, 35, 36, 38, 39, 40, 42, 44, 45, 46, 48, 49, 50, 51, 52, 54, 55, 56, 57, 58, 60, 62, 63, 64, 65, 66, 68, 69, 70, 72, 74, 75, 76, 77, 78, 80, 81, 82, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 98, 99, 100,

While prime numbers less than 100 are:
2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97

**Comment by appgurueu** [comment]
> > You should remove the isComposite function and instead make your changes to the PrimeCheck function.
> 
> What changes should I make? I am a bit clueless... I am just proposing an algorithm to check composite numbers.

Yes, but composite numbers are numbers that aren't prime, with the exception of one. You're thus duplicating primality sieve logic here. You should instead implement `isComposite` using `PrimeCheck` - or get rid of `isComposite` altogether - and improve the `PrimeCheck` function.

**Comment by tehliang** [comment]
> > > You should remove the isComposite function and instead make your changes to the PrimeCheck function.
> > 
> > 
> > What changes should I make? I am a bit clueless... I am just proposing an algorithm to check composite numbers.
> 
> Yes, but composite numbers are numbers that aren't prime, with the exception of one. You're thus duplicating primality sieve logic here. You should instead implement `isComposite` using `PrimeCheck` - or get rid of `isComposite` altogether - and improve the `PrimeCheck` function.

Hmm, I still do not quite understand. PrimeCheck is checking the prime number while the composite number is checking the reverse of it. How can I implement PrimeCheck to check whether the number is composite using the PrimeCheck algorithm?

**Comment by appgurueu** [comment]
> Hmm, I still do not quite understand. PrimeCheck checks the prime number while the IsComposite checks the composite number. Yes, what the algo doing was reverse of each other, but the code is not the same. How can I implement PrimeCheck to check whether the number is composite using the PrimeCheck algorithm?

```javascript
function isComposite(number) {
  return number > 1 && !PrimeCheck(number)
}
```

there you go.

**Review by appgurueu — APPROVED** [review]

**Pull request closed without merging**